wcコマンドの課題

ham-cap opened the pull request:
wcコマンドの課題を提出いたします。
ご確認のほどよろしくお願いいたします。

ham-cap committed:
引数にファイル名を指定した場合の値の受け取り部分をベタ書き（First commit）

wcコマンドは対象となるファイルが決まっている場合、引数に当該ファイル名を指定するため
引数があった場合の受け取り部分を書いた。
if文でARGVの有無を確認し、trueだった場合にこの処理を実行する。

ham-cap committed:
引数でファイルを指定した場合と標準入力（キーボード）から受け取った場合の処理をベタに書いた

行数、単語数、ファイルサイズのカウントと表示を実装
オプションは手つかず
なぜかパイプから受け取った場合の結果が違うため
標準入力から受け取った際の処理に問題があるかもしれない

ham-cap committed:
無駄な出力部分を削除

動作確認用の出力部分を削除した。
-lオプションを除いて、機能の実装は終了。
このバージョンを基にしてメソッド等を整理していく。

ham-cap committed:
ベタ書きをクラス・メソッドを使った書き方に修正し、-lオプションも実装した

引数にファイル名を指定した場合と他のコマンドからパイプで繋がれた場合、キーボードからの
標準入力があった場合の処理を実装。
ただし、複数個のファイル指定があった場合の処理ができていない。

ham-cap committed:
-lオプションを実装した

-lオプション使用時の出力メソッドを実装
引数が一つの場合、複数の場合、標準入力からの入力の場合に対応

ham-cap committed:
ターミナル上に表示する際の文字間隔を調整

行数、単語数、バイト数が表示される際の間隔を空白文字で調節していたが、
桁数によって表示がずれるため、rjustメソッドで右詰め表示になるようにした。

ham-cap committed:
行数、単語数、バイト数を数えるメソッドを微修正し、名前も変更

これらのメソッドの返り値を使って出力する際の記述が長くなりすぎるため
メソッド名を短いものに変更した。
また、出力時にしている共通の処理があったため、各メソッドに組み込むことで
出力時の記述を短くした。

yuuu commented on 06.wc/wc.rb:
```diff
@@ -0,0 +1,99 @@
+    end
+  end
+
+  def display_lines_and_name
```
`display_lines_and_name` と `display_lines_words_size_name` で大部分の処理が重複していますね。メソッドを共通化して、オプションで `-l` を指定できるようにした方がスッキリしそうなので検討をお願いします。

yuuu commented on 06.wc/wc.rb:
```diff
@@ -0,0 +1,99 @@
+  end
+
+  def count_lines
+    @inputed_line.count("\n").to_s.rjust(8)
```
ここで `rjust` していることで、totalを計算するときに `to_i` する手間が発生していますね。
`count_lines` では `Integer` を返すようにした方が後々使い勝手が良いと思います。出力フォーマットを整える処理は、出力時に行った方が保守性も良いです。

ham-cap replied:
この部分、実装時にも悩んだポイントなのでご相談させてください。
最初の段階では`count_lines`は単純に改行文字を数えて`Integer`で返すだけのメソッドで、`to_s`や`rjust`は出力部分でそれぞれの値に対して使っていました。
ただ、そうすると出力部分が長くなりすぎる場合があり、rubocopでも`offence`が出るため今のような書き方にしたという経緯があります。
例えば77行目は現状
```ruby
puts "#{@contents_of_files[0].count_lines}#{@contents_of_files[0].count_words}#{@contents_of_files[0].size} #{@contents_of_files[0].name}"
```
としていますが、`count_lines`、`count_words`、`size`の各メソッドから出力フォーマットを整えるための`.to_s.rjust(8)`の部分を切り出すと、
```ruby
puts "#{@contents_of_files[0].count_lines.to_s.rjust(8)}#{@contents_of_files[0].count_words.to_s.rjust(8)}#{@contents_of_files[0].size.to_s.rjust(8)} #{@contents_of_files[0].name}"`
```
となり、かなり長くなってrubocopでも`offence`がでます。
個人的には、おっしゃるとおり`.to_s.rjust(8)`の部分は出力部分に書いてしまいたいのですが、`rubocop`に指摘されている部分を放置するわけにもいかず、どうしたものかなと思っていました。
とはいえ、rubocopの`offence`にこだわりすぎるとかえって変なコードになってしまう気もするのですが。

また、`count_words`と`size`はそのままにして今回ご指摘いただいた`count_lines`だけ修正すればここまで長くはなりませんが、そうなると各メソッドで出力フォーマットを整える処理を含むものとそうでないものができてしまい、ちょっと気持ち悪いかなと思うのですが、ここはそこまでこだわらなくてもよいのでしょうか。

yuuu replied:
確かにこれは長すぎますね。

```ruby
puts "#{@contents_of_files[0].count_lines.to_s.rjust(8)}#{@contents_of_files[0].count_words.to_s.rjust(8)}#{@contents_of_files[0].size.to_s.rjust(8)} #{@contents_of_files[0].name}"
```

rubocopの指摘が `Style/LineLength` である前提で回答します。違ったらすみません。

短くする方法はいろいろあるのですが、一例を紹介しますね。
まず `contents_of_files[0]` というコードが重複しているので、一旦短い名前の別の変数に代入しても良いと思います。

```ruby
file = @contents_of_files[0]
puts "#{file.count_lines.to_s.rjust(8)}#{file.count_words.to_s.rjust(8)}#{file.size.to_s.rjust(8)} #{file.name}"
```

すべての計算を1行で行う必要はないので、予めフォーマットしておいて別の変数に代入することもできます。

```ruby
file = @contents_of_files[0]
lines = file.count_lines.to_s.rjust(8)
words = file.count_words.to_s.rjust(8)
chars = file.size.to_s.rjust(8)
puts "#{lines}#{words}#{chars} #{file.name}"
```

また、`to_s.rjust(8)` も重複しているのでまとめてしまいましょう。

```ruby
file = @contents_of_files[0]
lines, words, chars = [file.count_lines, file.count_words, file.size].map(&:to_s).map { |c| c.rjust(8) }
puts "#{lines}#{words}#{chars} #{file.name}"
```

`lines`, `words`, `chars` は単純につなげているだけなので、`join` するとさらに短くなりますね。

```ruby
file = @contents_of_files[0]
counts = [file.count_lines, file.count_words, file.size].map(&:to_s).map { |c| c.rjust(8) }
puts "#{counts.join} #{file.name}"
```

これでもまだ `file` が重複しているのでもっとリファクタリングすることはできますが、rubocopの指摘はクリアできるのではないかと思います。

---

ちなみに、今回の場合はコードの見直しでこれだけ短くなりましたが、どうしても短くならない場合は `puts` ではなく `print` で部分的に出力していくのも手だと思います。

yuuu commented on 06.wc/wc.rb:
```diff
@@ -0,0 +1,99 @@
+      @line_numbers = []
+      @contents_of_files.each do |content|
+        puts "#{content.count_lines} #{content.name}"
+        @line_numbers << content.count_lines.to_i
```
`count_lines` で `Integer` を返すようにすれば、ここはeachの外に出して、以下のように書いてしまって良いと思います。

```ruby
puts "#{@contents_of_files.map(&:count_lines).sum.to_s.rjust(8)} total"
```

`@line_numbers` が不要になりますね。

yuuu commented on 06.wc/wc.rb:
```diff
@@ -0,0 +1,99 @@
+  def initialize
+    @contents_of_files = []
+    if ARGV.size >= 1
+      ARGV.each do |file_name|
```
`Array` から `Array` を作る処理はl `Array#map` を使ったほうがスッキリ書けますね。

```ruby
@contents_of_files = ARGV.map { |file_name| ContentsOfFiles.new(file_name) }
```

ham-cap committed:
レビューの指摘事項を修正

・オプションの有無によって出力用のメソッドを分けて定義していたが、重複する処理が多かったため共通化した
・eachで繰り返し処理を行っていた一部の処理をmapを用いたものに変更した
・行数等をカウントするメソッドからto_sとrjustの処理を切り離し、出力時に行うように変更した

ham-cap commented on 06.wc/wc.rb:
```diff
@@ -31,69 +36,56 @@ def initialize
 
-  def display_lines_and_name
-    if @contents_of_files.size == 1
+  def display
```
`display_lines_and_name`と`display_lines_words_size_name`を1つのメソッドにまとめて`display`として定義し直しました。一応、動作自体に問題はありませんが、rubocopのoffenceで`Metrics/MethodLength`と`Metrics/PerceivedComplexity`が出てしまいました。
メソッドの行数としては`26/20`で6行超過、Complexityは`23/8`と大幅に超過しています。
たしかにif分の分岐条件が複雑になってしまっており、それぞれの処理ももう少しシンプルに書けそうな気がしていろいろ試してみたのですが、なかなか上手くいかないため、アドバイスいただければ幸いです。

yuuu replied:
共通化の検討、ありがとうございます。確かに、かなり処理が複雑化していますね。
私の当初のイメージは `display_lines_words_size_name` の処理をベースにして 「 `-l` が存在しない場合のみ `word_numbers` と `bytes` を出力する」ようにすればコードもスッキリするだろうと見込んでいました。この方針であればおそらく今の状態よりは複雑度が下がると思います。

ただ、Rubocopをクリアできるかどうかは少し怪しいと思っているので、必要に応じてサブメソッド化しないと駄目かもしれませんね。

ham-cap replied:
コメントありがとうございます。

「`-l`が存在しない場合のみ`word_numbers`と`bytes`を出力する」という理屈自体は理解できるのですが、それを実装するための具体的なロジックがうまくイメージできないため、いったん出力時の処理について整理してみました。
まず、このプログラムの入力は3パターンあり、それぞれに対応する出力の仕方が違うと思います。
1. 引数が単数の場合→行数、単語数、バイト数、ファイル名を表示
1. 引数が複数の場合→行数、単語数、バイト数、ファイル名、各項目の合計を表示
1. 標準入力の場合→行数、単語数、バイト数を表示
更にこれらを`-l`の有無によって分岐させると1〜3のパターンの入力に対して表示項目を減らしたバリエーションが一つずつ増えるので、計6通りの出力になります。
したがって、一つのメソッドにまとめると今回のようにif文の分岐が6つできてしまい、しかも同じ条件式で引数の数と`-l`の有無を判定するため複雑になってしまうのだと思います。

当初、メソッドを`display_lines_words_size_name`と`display_lines_and_name`に分けていたのは、6通りの分岐はちょっと多すぎるなと考えたからです。トップレベルのロジックの段階で`-l`の有無を判定したうえで実行するメソッドを選択することで一つのメソッド内での分岐を少なくすることと、`&&`でつなぐような条件にしなくて済むようにと考えて別のメソッドとして定義したという経緯があります。
ただ、これに関しては`-l`の有無の判定方法を工夫することで解決できそうで、もしかするとそれが@yuuu さんのおっしゃっている「`-l`が存在しない場合のみ`word_numbers`と`bytes`を出力する」につながるのかなと思うのですが、具体的にどんな方法がありますでしょうか。自分で考えていても現状の`if @contents_of_files.size == 1 && @options`のように`&&`でつなぐ形で判定する方法しか思いつかないので。

また、サブメソッドとはどういうものかご教示いただけますでしょうか。
調べてみてもそれらしい情報にいきあたらないため、具体的にどういうものをサブメソッドと呼ぶのか教えていただけると助かります。

ham-cap committed:
共通化したメソッドを再度2つに分割した

displayメソッドが複雑化しすぎたため

ham-cap committed:
メソッド内の一部の処理を簡略化及び別メソッドへの切り出し等を行った

・出力の際の表示位置調整のため、to_sとrjustの組み合わせで加工することが多いため、Integerクラスにメソッドとして定義した
・出力用メソッドの名前が判別しにくかったため、display_lines_words_size_nameをdisplay_multi_statusへ変更した
・display_multi_status内の処理がrubocopに複雑すぎると指摘されたため、ファイル名が複数わたってきた場合にtotalを表示する処理を別メソッドへ切り出した
